packages/orm/src/types.ts

```diff
@@ -29,10 +35,70 @@ export enum Relations {
   ManyToMany = "manyToMany"
 }
 
+/**
+ * An utility type, created for two purposes:
+ *  1. Checks if each model have defined modelName field with a string literal type of that value.
+ *  2. Checks if the modelName value matches with the key under which we add a model to the Schema type.
+ * 
+ * Handling both cases preserves from possible typo's or modelName duplicates that can be caught by static checks.
+ * Another value it brings, is a naming consistency.
+ */
+export type ValidateSchema<Schema extends ModelClassMap> = { [K in keyof Schema]: ModelName<Schema[K]> extends K ? Schema[K] : never };
+
 /**
  * Returns the type of the passed model class.
  */
-export type ModelClassType<M extends AnyModel> = ReturnType<M["getClass"]>;
+export type ModelClassType<MClass extends AnyModel> = ReturnType<MClass["getClass"]>;
+
+/**
+ * Returns the type of static modelName field. This field should be described by a string literal type.
+ */
+export type ModelName<MClassType extends typeof AnyModel> = MClassType['modelName'];
+
+/**
+ * Extracts all non-serializable types from the model's fields.
+ * 
+ * It is being used to determine to which model types, the relationship can be created.
+ */
+export type ModelClassTypeFromModelFields<MClass extends AnyModel, MFields extends Required<ModelFields<MClass>> = Required<ModelFields<MClass>>> = 
```

Comment:
Currently, the type checks in decorators are not working perfectly due to technical reasons, I have one idea how I could improve that but it should not block this PR from merging. I will try to explain this in more details on Tuesdays sync.